dozer-ingestion/src/utils.rs

```diff
@@ -0,0 +1,21 @@
+#[macro_export]
+macro_rules! retry_on_network_failure {
```

Comment:
While I appreciate the effort of trying to reuse this piece of code everywhere, I'm a little uneasy that such a complicated macro is used extensively. It took me half an hour to understand what it does!

Per my understanding, this macro only has to be a macro instead of a function because the usage in `exec_iter_impl`, where we pass `continue 'main` as the `reconnect`, which is not something a function can do.

If this understanding is correct, how about we change the macro to a function and write a similar macro in `exec_iter_impl` just for being used in that function?

This macro can be changed to a function like:

```rust
/// Connection network failure should be handled in `reconnect`
async fn retry_on_network_failure<
    Operation,
    OperationFuture,
    ErrorPredicate,
    Reconnect,
    ReconnectFuture,
    Output,
    Error
>(
    description: &'static str,
    mut operation: Operation,
    network_error_predicate: ErrorPredicate,
    mut reconnect: Option<Reconnect>
) -> Result<Output, Error> 
where
    Operation: FnMut() -> OperationFuture,
    OperationFuture: Future<Output = Result<Output, Error>>,
    ErrorPredicate: FnMut(&Error) -> bool,
    Reconnect: FnMut() -> ReconnectFuture,
    ReconnectFuture: Future<Output = Result<(), Error>>,
{
    ...
}
```

Reply:
Managing the lifetime of `self` in futures is complicated, and with `loop` it gets really frustrating.

Reply:
So this function signature doesn't work? I was thinking we can get away from that with all the `Operation`, `OperationFuture`, ... generic types.

Reply:
Having trouble with `captured variable cannot escape FnMut closure body` with `self.reconnect()` in 
```rust
retry_on_network_failure(
            "prepare",
            || self.inner.prepare(query),
            is_network_failure,
            Some(|| self.reconnect())
```

Reply:
Let me play with it for a little. If it's a deep rabbit hole we'll just leave it there.

Reply:
OK this is indeed deep. https://smallcultfollowing.com/babysteps/blog/2023/03/29/thoughts-on-async-closures/. I give up.

Reply:
I sure learned a few lessons this week about rust lifetime management, self-referential structs, and loops. I spent a lot of time solving and dodging these lifetime issues.

Reply:
That's only natural for every Rust programmer